freebsd/utils/jail.cpp

```diff
@@ -0,0 +1,323 @@
+#include "utils/process/operations.hpp"
+#include "utils/process/status.hpp"
+
+namespace process = utils::process;
+namespace fs = utils::fs;
+
+using utils::process::args_vector;
+using utils::process::child;
+
+
+static const int jail_name_max_len = MAXHOSTNAMELEN - 1;
+static const char* jail_name_prefix = "kyua";
+
+
+/// Functor to run a program.
+class run {
+    /// Program binary absolute path.
+    const utils::fs::path& _program;
+
+    /// Program arguments.
+    const args_vector& _args;
+
+public:
+    /// Constructor.
+    ///
+    /// \param program Program binary absolute path.
+    /// \param args Program arguments.
+    run(
+        const utils::fs::path& program,
+        const args_vector& args) :
+        _program(program),
+        _args(args)
+    {
+    }
+
+    /// Body of the subprocess.
+    void
+    operator()(void)
+    {
+        process::exec(_program, _args);
+    }
+};
+
+
+namespace freebsd {
+namespace utils {
+
+
+std::vector< std::string >
+jail::parse_params_string(const std::string& str)
+{
+    std::vector< std::string > params;
+    std::string p;
+    char quote = 0;
+
+    for (const char& c : str) {
+        // whitespace delimited parameter
+        if (quote == 0) {
+            if (std::isspace(c)) {
+                if (p.empty())
+                    continue;
+                params.push_back(p);
+                p = "";
+            }
+            else if (c == '"' || c == '\'') {
+                if (!p.empty())
+                    params.push_back(p);
+                p = "";
+                quote = c;
+            }
+            else
+                p += c;
+        }
+
+        // quoted parameter
+        else {
+            if (c == quote) {
+                if (!p.empty())
+                    params.push_back(p);
+                p = "";
+                quote = 0;
+            }
+            else
+                p += c;
+        }
+    }
+
+    // leftovers
+    if (!p.empty())
+        params.push_back(p);
+
+    return params;
+}
+
+
+/// Constructs a jail name based on program and test case.
+///
+/// The formula is "kyua" + <program path> + "_" + <test case name>.
+/// All non-alphanumeric chars are replaced with "_".
+///
+/// If a resulting string exceeds maximum allowed length of a jail name,
+/// then it's shortened from the left side keeping the "kyua" prefix.
+///
+/// \param program The test program.
+/// \param test_case_name Name of the test case.
+///
+/// \return A jail name string.
+std::string
+jail::make_name(const fs::path& program,
+                const std::string& test_case_name)
+{
+    std::string name = std::regex_replace(
+        program.str() + "_" + test_case_name,
+        std::regex(R"([^A-Za-z0-9_])"),
+        "_");
+
+    const std::string::size_type limit =
+        jail_name_max_len - strlen(jail_name_prefix);
+    if (name.length() > limit)
+        name.erase(0, name.length() - limit);
+
+    return jail_name_prefix + name;
```

Comment:
_(picking a line)_
Does this code function properly if the jail already exists, or will it fail (repeatedly) with an obscure command line error?

Reply:
We had a discussion with markj@ about the possible clashing of temporary jail naming, with ideas with hashing. The decision was postponed for the future, considering the low chance of clashing for today's FreeBSD Test Suite. What do you think about current conclusion?